File: builder-pattern/java/src/PhoneBuilder.java
package src;

// 手机建造者实现了建造者接口
public class PhoneBuilder implements Builder {
  private Phone phone;

  public void reset() {
    this.phone = new Phone();
  }

  public Builder setName(String name) {
    this.phone.setName(name);
    return this;
  }

  public Builder setScreen(Integer[] screen) {
    this.phone.setScreen(screen);
    return this;
  }

  public Builder setGPU(Integer no) {
    this.phone.setGpuType(no);
    return this;
  }

  public Phone getProduct() {
    return this.phone;
  }
}

File: builder-pattern/go/src/PhoneBuilder.go
package src

type PhoneBuilder struct {
  phone *Phone
}

func (p *PhoneBuilder) Reset() {
  p.phone = new(Phone)
}

func (p *PhoneBuilder) SetName(name string) Builder {
  p.phone.SetName(name)
  return (Builder)(p)
}

func (p *PhoneBuilder) SetScreen(screen []int) Builder {
  p.phone.SetScreen(screen)
  return (Builder)(p)
}

func (p *PhoneBuilder) SetGPU(no int) Builder {
  p.phone.SetGpuType(no)
  return (Builder)(p)
}

func (p *PhoneBuilder) GetProduct() *Phone {
  return p.phone
}
